design/debug-collector.md

```diff
@@ -0,0 +1,68 @@
+
+### Choice of backend
+
+The collector will access memory on the target device either by using the probe-rs Rust library, or by talking to a gdb server that is attached to the target. This choice will be given to the user of the CLI, using `--attach` for probe-rs, or `--gdb-host` to use gdb.
+
+#### probe-rs
+
+probe-rs provides functions to attach to chips over a usb connection, as well as to read from or write to memory locations.
+
+To use this option, users must specify the type of chip that they wish to connect to, which is required by probe-rs: `--attach stm32`.
+
+#### gdb
+
+If the user already has a GDB server that attaches to the chip set up (e.g. OpenOCD), or would rather use such a tool to attach to the target, then they can specify the address of the gdb server that they are running: `--gbd-host localhost:3333`.
+
+In this case, the debug-collector will communicate to the given gdb server over TCP in order to read and write to memory.
+
+### repr(C)
+
+Rust's ABI is undefined, so the representation of structures in memory is unpredictable across compiler versions and architectures. To allow the debug-collector to understand the raw binary it reads from the chip, the ekotrace log data structures must have the `repr(C)` attribute, which tells the compiler to use C's representation of structs in binary. To find the fields of `repr(C)` structs, the debug-collector uses constant offsets for each given word size. This method is fragile to implementation changes to the structs that the debug-collector is accessing.
+
+We would have to add this attribute to `DynamicHistory` and `ChunkedReportState` in ekotrace as well as to `FixedSliceVec` in fixed-slice-vec.
+
+## Symbol resolution
+
+For each tracer, the user will either pass in the memory location of a pointer to the tracer, or will pass in the name of the symbol of the pointer to the tracer. If the latter is used, then the user would also need to pass in an ELF file containing the symbol table of the program running on the target device. The debug-collector will automatically find the memory location of the given symbol by looping through the given symbol table.
+
+## Locking
+
+The debug-collector reads and writes to the ekotrace logs that the tracers running on the device have access to, so the debug-collector must ensure that it is not accessing logs at the same time as tracers. To do so, we can add an `in_use: Atomic<bool>` field to the DynamicHistory struct that will be set to true while the tracer is accessing the log. When the debug-collector pulls the logs, it will halt the CPU of the target and then check the value of `in_use`, aborting if true. If `in_use` is false, then the debug-collector can safely access the logs and then unhalt the CPU when finished. The use of atomic CPU instructions by the `Atomic<bool>` type should prevent race conditions.
+
```

Comment:
I think it'd be helpful if you could describe how we intend on performing the "atomic" operation here from the perspective of the debugger.

When the tool/debugger goes to set the `in_use` variable to `true/false`, what instructions/operations will be performed, is it a compare-and-swap/etc?

Reply:
By tool/debugger, do you mean the debug-collector itself? I was originally proposing that it would halt the CPU in order to prevent tracers from entering the mutually exclusive area, so it wouldn't write to the in_use variable at all. It would just check in_use to make sure that when the CPU is halted, the tracer is not already in the middle of a log access.

Reply:
Ah I see, I misread and didn't catch the halt-and-read part.